### DISCUSSION TITLE
How to optimize antenna position using sionna RT

### DISCUSSION THREAD

**Discussion by JackYifan** [NVlabs/sionna#435] [Q&A]
Hello, I would like to ask how to optimize antenna position using sionna RT？
```python
# Configure an SGD optimizer
optimizer = tf.keras.optimizers.RMSprop(1e-3)

# Number of training steps
num_steps = 30
scene.transmitters['tx'].position = tf.Variable([0.0,0.0,2.5],trainable=True)
def train_step():
    """A single training step"""
    with tf.GradientTape() as tape:
        # Compute coverage of the target area
        target_cm = scene.coverage_map(cm_center=target_center,
                                       cm_orientation=target_orientation, 
                                       cm_size=target_size, # Target area 
                                       cm_cell_size=cm_cell_size,
                                       diffraction=True, scattering=True, # Enable diffraction and scattering in addition to reflection and LoS
                                       check_scene=False) # Don't check the scene prior to compute to speed things up
        scaling = 1e6
        rate = tf.reduce_mean(tf.math.log(1. + target_cm.as_tensor()*scaling))/tf.math.log(2.)
        loss = -rate
    
    # Compute gradients and apply through the optimizer
    grads = tape.gradient(loss, tape.watched_variables())
    optimizer.apply_gradients(zip(grads, tape.watched_variables()))
    return rate

for step in range(num_steps):       
    rate = train_step()
    print(f"Training step {step} - Rate: {rate.numpy():.2E} bit - tx position: {scene.transmitters['tx'].position.numpy()}", end='\n')
```
I tried but get the error "No gradients provided for any variable".
```
No gradients provided for any variable: (['Variable:0'],). Provided `grads_and_vars` is ((None, <tf.Variable 'Variable:0' shape=(3,) dtype=float32, numpy=array([0. , 0. , 2.5], dtype=float32)>),).
```

**Comment by SebastianCa** [comment]
Hi @JackYifan,

please see #440 for a detailed explanation. The short version is that SionnaRT simply does not provide gradients w.r.t. the transmitter position.

**Comment by merlinND** [comment]
Hello @JackYifan,

Please refrain from opening multiple issues / discussion about the same question, as it increases the workload for maintainers.
Closing as duplicate of #440.
